bigchaindb/upsert_validator/validator_election.py

```diff
@@ -0,0 +1,139 @@
+
+        self.operation = self.VALIDATOR_ELECTION
+
+    @classmethod
+    def current_validators(cls, bigchain):
+        """Return a dictionary of validators with key as `public_key` and
+           value as the `voting_power`
+        """
+
+        validators = {}
+        for validator in bigchain.get_validators():
+            # NOTE: we assume that Tendermint encodes public key in base64
+            public_key = public_key_from_ed25519_key(key_from_base64(validator['pub_key']['value']))
+            validators[public_key] = validator['voting_power']
+
+        return validators
+
+    @classmethod
+    def recipients(cls, bigchain):
+        """Convert validator dictionary to a recipient list for `Transaction`"""
+
+        recipients = []
+        for public_key, voting_power in cls.current_validators(bigchain).items():
+            recipients.append(([public_key], voting_power))
+
+        return recipients
+
+    @classmethod
+    def is_same_topology(cls, current_topology, election_topology):
+        voters = {}
+        for voter in election_topology:
+            if len(voter.public_keys) > 1:
+                return False
+
+            [public_key] = voter.public_keys
+            voting_power = voter.amount
+            voters[public_key] = voting_power
+
+        # Check whether the voters and their votes is same to that of the
+        # validators and their voting power in the network
+        return (current_topology == voters)
+
+    def validate(self, bigchain, current_transactions=[]):
+        """Validate election transaction
+        NOTE:
+        * A valid election is initiated by an existing validator.
+
+        * A valid election is one where voters are validators and votes are
+          alloacted according to the voting power of each validator node.
+
+        Args:
+            bigchain (BigchainDB): an instantiated bigchaindb.tendermint.BigchainDB object.
+
+        Returns:
+            The transaction (Transaction) if the transaction is valid else it
+            raises an exception describing the reason why the transaction is
+            invalid.
+
+        Raises:
+            ValidationError: If the transaction is invalid
+        """
+        input_conditions = []
+
+        duplicates = any(txn for txn in current_transactions if txn.id == self.id)
+        if bigchain.get_transaction(self.id, cls=ValidatorElection) or duplicates:
+            raise DuplicateTransaction('transaction `{}` already exists'
+                                       .format(self.id))
+
+        if not self.inputs_valid(input_conditions):
+            raise InvalidSignature('Transaction signature is invalid.')
+
+        current_validators = self.current_validators(bigchain)
+
+        # NOTE: Proposer should be a single node
+        if len(self.inputs) != 1:
+            raise TypeError('`tx_signers` must be a list instance or length one')
```

Comment:
It makes sense to have a specifc exception in the case like this so that a caller can easily distinguish exceptions.

Reply:
You mean a custom exception?

Reply:
Yes.